"Do" block parsing fixed to dereference expression tuples.

MatthewHambley opened the pull request:
When parsing a line such as `do i = 1, size(this%array)` fparser1 returns something like `DO i = 1, size(F2PY_EXPR_TUPLE_9)`.

This branch introduces a unit test to exercise the "do" parser, including this fault condition. It then fixes the issue by resolving the "expression tuple" string to the correct value. This is done after pattern matching is performed against the string, which was the reason the mapping was introduced in the first place.

While doing this I noticed that each of the tests in "test_block_stmts.py" imported a common set of bits. I've hoisted this up to module level to save typing.

MatthewHambley commented:
> Please could you extend the new tests that you've implemented to fully cover R826 (and sub-rules) for block DO constructs.

Do you mean R826?

In my copy of the standard (BS ISO/IEC 1539-1:2010) that appears in the middle of section 8.1.6.3 "Form of the nonblock DO construct". Do you instead mean R814 which is the start of the section 8.1.6.2 "Form of the block DO construct"?

arporter commented:
I think I do. However, I suspect we're not looking at the same standard as for me R826 appears at the start of 8.1.6.1.1 (Form of the block Do construct). The document I'm looking at is the one pointed to from the fparser docs and is the 2004 draft of the 2003 spec. Its reference no. is ISO/IEC JTC1/SC22/WG5 N1601. Which one are you looking at?

MatthewHambley commented:
I'm looking at BS ISO/IEC 1539-1:2010 which is some 6 years more recent than yours. I've implemented a comprehensive suite of tests for the "DO" block as given in the section "Form of the block DO construct". There are a lot of permutations so it adds considerably to the time taken to perform the testing.

It has also thrown up a number of nasty little gremlins which I have also squashed.

Finally I flipped the sense of an assertion in `src/fparser/one/tests/test_block_stmts.py` which I believe was in error but not failing due to some other issue. In particular, the test is for a full set of "implicit" rules so I don't think we should be expecting a comment telling us that default rules are in operation.

Something has gone awry with the python 2 run which is why the integration tests failed. It passes for Python 3. I wonder if some timeout is tripping and killing the process at 20 minutes, or around 80% of the way through.

rupertford commented:
I think that the standard @MatthewHambley has been looking at is Fortran2008 whereas the one @arporter has been suggesting is Fortran2003.

For info, the best (free) place to look for the drafts is here  https://wg5-fortran.org/ and then pick the appropriate version of Fortran from the drop down menu.

However, seeing as fparser1 was only supporting Fortran up to Fortran 90 then it's not clear which standard should be picked. Perhaps going with Matthews 2008 version would minimise hassle and then he could point out any additional syntax (from Fortran90 to 2008) under the 'Support for Fortran Standards beyond Fortran90' documentation section if there is any, and he could add a comment to the classes and tests to explain that they respect the 2008 standard and its numbering rules?

rupertford commented:
I've just looked at the travis logs and whilst I'm not reviewing this PR I feel I need to point out that the number of tests added is excessive to put it mildly.

Yes, in theory we could test every option in every combination, but in practice this is completely impractical, as evidenced by the increase in time required to do the tests (from under a minute to over 20 minutes), just by adding the tests for a small number of rules. Imagine if this approach was taken for all rules.

The way I suggest the tests are done is to test each option in isolation and only in combination when one option is only valid if another option exists (and then test for failure when this condition is not met).

What we look for is 100% line coverage at the minimum with additional combination tests where it makes sense i.e. where something could go wrong in the code logic.

arporter commented on src/fparser/one/parsefortran.py:
```diff
@@ -130,16 +129,16 @@ def parse(self):
                                           'while processing line',
                                           reader.linecount, reader.linecount)
-                logging.getLogger(__name__).critical(message)
+                logger.critical(message)
```
Bug fix?

MatthewHambley replied:
Just a tidy up. It avoids making the `logging.getLogger(__name__)` call multiple times in the same function.

arporter commented on src/fparser/one/tests/test_do.py:
```diff
@@ -0,0 +1,212 @@
+# -*- coding: utf-8 -*-
```
Please name this file with the rule name and number (see item 6. of the review guidelines).

MatthewHambley replied:
Item 6 seems to refer to the Mk.II parser and Fortran 2003 syntax. But since you asked, I have followed the same convention for this one.

arporter replied:
Great, thanks.

arporter commented on src/fparser/one/tests/test_do.py:
```diff
@@ -0,0 +1,212 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+'''
+Tests all permutations of the DO block. There are a lot of them.
+'''
```
No kidding, we've got 190,000 tests!

arporter commented on src/fparser/one/tests/test_do.py:
```diff
@@ -0,0 +1,212 @@
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('end_name', [None, 'loop_name', 'wrong_name'])
+@pytest.mark.parametrize('end_label', [None, '123', '456'])
+def test_do(name, label, control_comma, initial_expression,
```
Please name these tests for the rules they are checking and/or document this in the docstrings.

arporter commented on src/fparser/one/tests/test_do.py:
```diff
@@ -0,0 +1,212 @@
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('end_name', [None, 'loop_name', 'wrong_name'])
+@pytest.mark.parametrize('end_label', [None, '123', '456'])
+def test_do_concurrent(name, label, control_comma, typespec,
```
Is 'do concurrent' F2008? If so, please document this in the fparser1 user guide.

MatthewHambley replied:
It is and as such I have removed it from the testing. This seemed appropriate as there is no guarantee that fparser-1 will support anything beyond Fortran 95. As it happens it will support this because it treats everything after "do" as a plain string but since you are keen to cut down the amount of testing...

arporter commented:
Back to @MatthewHambley.

MatthewHambley commented:
Passing back for further review. I hope I have addressed all previous concerns.

arporter commented on src/fparser/one/tests/test_do_block_r814.py:
```diff
@@ -48,43 +52,43 @@
     comma_snippet = ', ' if control_comma else None
-    inc_snippet = ', ' + incrament_expression if incrament_expression else None
-    do_code = '''{name}do {label}{comma}variable = {init}, {term}{inc}
+    # TODO: Although the Fortran standard allows for "continue" to be used in
```
Please just change this to comment to say that fparser1 doesn't support it and fparser1 is being deprecated. (Otherwise we have to have an issue associated with the TODO and nobody wants that...)

arporter commented on src/fparser/one/tests/test_do_block_r814.py:
```diff
@@ -48,43 +52,43 @@
+    Checks that the "do" loop parser understands the "for-next" variant of
+    the syntax. This is defined in BS ISO/IEC 1539-1:2010 with R814-R822.
+
+    TODO: Only the terminal expression is tested. This is a short-cut and
```
Please also turn this into just a normal comment otherwise we have to have a new Issue and we aren't actively developing fparser1.

arporter replied:
On second thoughts, don't worry.

arporter commented:
We're back down to 1740 tests that run in just under 10s so that's OK. Coverage of the diff is still 100%.
Testing for the F2003 'do concurrent' has been dropped so no need to update the documentation.
Documentation builds fine.
